api/app/src/main/kotlin/packit/controllers/PacketController.kt

```diff
@@ -77,4 +81,29 @@ class PacketController(
             .headers(response.second)
             .body(response.first)
     }
+
+    @GetMapping("/{name}/{id}/zip")
+    @PreAuthorize("@authz.canReadPacket(#root, #id, #name)")
+    // TODO: Verify authorization for accessing specific files: whether user has access to all files, or artefacts only.
```

Comment:
Most importantly, you need to check that the hashes that have been specified really belong to the packet in question. The current implementation does access checks on `id` and `name`, but then doesn't really use these values.

Nothing prevents a user from specifying a packet ID that they do have access to, but with a list of hashes from a different packet for which they do not have access. The API server will happily return those files.

This is actually a bug in `findFile` and will need to be fixed there as well (I also don't understand why `findFile` uses `canReadPacketMetadata` and not `canReadPacket`).

Reply:
I've fixed the `findFile` bug in https://github.com/mrc-ide/packit/pull/171

> I also don't understand why findFile uses canReadPacketMetadata and not canReadPacket

It turned out `canReadPacket` and `canReadPacketMetadata` basically do the same thing, but the existing `canReadPacket` was very error prone (in your code, the user controls `name` and can use it to access any ID if they have permission for at least one packet). I've tried to clarify the interface. For this endpoint, after my PR is merged, you should be able to use:
```kotlin
@PreAuthorize("@authz.canReadPacket(#root, #id")` 
```